orttraining/orttraining/training_api/module.cc

```diff
@@ -2,12 +2,137 @@
 // Licensed under the MIT License.
 
 #include "core/session/inference_session.h"
+#include "core/session/environment.h"
+
 #include "orttraining/training_api/include/module.h"
+#include "orttraining/training_api/include/utils.h"
+
+using namespace onnxruntime;
 
 namespace onnxruntime {
 namespace training {
 namespace api {
 
+Status Parameter::AllocateGrad(const std::string& gradient_name, const SessionState& sess_state) {
+  // assert param is allocated
+  ORT_ENFORCE(data_.IsAllocated());
+  ORT_ENFORCE(requires_grad_);
+  gradient_name_ = gradient_name;
+  ORT_ENFORCE(OrtValueLike(sess_state, data_, gradient_).IsOK());
+  return Status::OK();
+}
+
+Status Parameter::ResetGrad() {
+  //TODO: make use of lazy_reset_grad input instead 
+  Tensor* p_tensor = gradient_.GetMutable<Tensor>();
+  const auto& device = p_tensor->Location().device;
+  if (device.Type() == OrtDevice::CPU) {
+    memset(p_tensor->MutableDataRaw(), 0, p_tensor->SizeInBytes());
+  }
+#if defined(USE_CUDA) || defined(USE_ROCM)
+  else if (device.Type() == OrtDevice::GPU) {
+    ORT_NOT_IMPLEMENTED("Not implemented.");
+  }
+#endif
+  else {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Unknown device type ", device.Type(), " for param:", name_);
+  }
+  return Status::OK();
+}
+Module::Module(const std::string& train_model_path_or_bytes,
+               std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters,
+               const std::optional<std::string>& eval_model_path_or_bytes) {
+  // create value copy as same params are passed to Optimizer constructor
+  parameters_ = parameters;
+
+  auto so = onnxruntime::SessionOptions();
+  std::unique_ptr<Environment> env;
```

Comment:
Since both module and optimizer are within the same process, they should share the env for sessions. We should find out the consequences of 2 envs.

Reply:
yes added a todo, will come together with  sharing threadpool